src/linear_algebra/affine_space.lean

```diff
@@ -402,6 +468,37 @@ begin
     exact finset.eq_zero_of_sum_eq_zero hw h2b i hi }
 end
 
+variables {k V}
+
+/-- If a family is affinely independent, so is any subfamily given by
+composition of an embedding into index type with the original
+family. -/
+lemma affine_independent_embedding_of_affine_independent {ι2 : Type*} (f : ι2 ↪ ι) {p : ι → P}
+    (ha : affine_independent k V p) : affine_independent k V (p ∘ f) :=
+begin
+  intros fs w hw hs i0 hi0,
+  let fs' := fs.map f,
+  let w' := λ i, if h : ∃ i2, f i2 = i then w h.some else 0,
+  have hw' : ∀ i2 : ι2, w' (f i2) = w i2,
+  { intro i2,
+    have h : ∃ i : ι2, f i = f i2 := ⟨i2, rfl⟩,
+    have hs : h.some = i2 := f.injective h.some_spec,
+    simp_rw [w', dif_pos h, hs] },
+  have hw's : ∑ i in fs', w' i = 0,
+  { rw [←hw, finset.sum_map],
+    simp [hw'] },
+  have hs' : fs'.weighted_vsub V p w' = 0,
+  { rw [←hs, finset.weighted_vsub_apply, finset.weighted_vsub_apply, finset.sum_map],
+    simp [hw'] },
+  rw [←ha fs' w' hw's hs' (f i0) ((finset.mem_map' _).2 hi0), hw']
+end
```

Comment:
I've found when working with affine combinations that there are lots of such arguments, moving from sums over one type or subset to another, that are mathematically trivial, but where formalisation needs to get into the details of the exact functions involved each time. Sometimes the sums use a subtype, sometimes another embedding as here. (At some point I could well run into lean#359 and need to duplicate things for `fin n` that already exist for subtypes.) Sometimes the manipulations of functions naturally involve `set.indicator`, sometimes `function.update`, sometimes they naturally involve an ad hoc (possibly dependent) `if`/`then` as here, or combine multiple kinds of manipulations of functions. Where there appear to be general lemmas not extremely specific to the combination of functions in a particular case, I've added them to `big_operators` (but that doesn't tend to shorten the proofs much, as you may still need a few lines to convert things into the exact form needed by the general lemmas, and that's replacing a proof that's only a few lines anyway). There might be some scope for a general "obvious equalities of sums and products" tactic, but (a) I don't know how to specify such a tactic that would both be useful and implementable, (b) I'm not sure it would help here, when the manipulation of sums is already done with `sum_map` together with `simp` and (c) a general tactic might well not be able to handle the sort of ad hoc function definition involved here.

Reply:
Thank you for the long, thought-out response. If we built some solution that shortens the proofs drastically, it would be out of scope for this PR.